src/irr/core/LRUcache.h

```diff
@@ -63,65 +50,79 @@ class LRUcache
 				m_list.popBack();
 			}
-			m_list.pushFront(std::pair<Key, Value>(k, v));
+			m_list.pushFront(std::make_pair(std::forward<K>(k), std::forward<V>(v)));
 			m_shortcut_map[k] = m_list.getFirstAddress();
 		}
 		else
 		{
-			m_list.get(iterator->second)->data = std::pair<Key, Value>(k, v);
+			m_list.get(iterator->second)->data.second = std::forward<V>(v);
 			m_list.moveToFront(iterator->second);
 		}
-
 	}
-	inline void insert(Key& k, Value&& v)
+public:
+	//Constructor
+	inline LRUcache(const uint32_t capacity) : m_capacity(capacity), m_shortcut_map(), m_list(capacity)
 	{
-		auto iterator = m_shortcut_map.find(k);
-		if (iterator == m_shortcut_map.end())
-		{
-			if (m_shortcut_map.size() >= m_capacity)
-			{
-				m_shortcut_map.erase(m_list.getBack()->data.first);
-				m_list.popBack();
-			}
-			m_list.pushFront(std::pair<Key, Value>(k, std::move(v)));
-			m_shortcut_map[k] = m_list.getFirstAddress();
+		assert(capacity > 1);
+		m_shortcut_map.reserve(capacity);
+	}
 
-		}
-		else
+#ifdef _IRR_DEBUG
+	inline void print()
+	{
+		auto node = m_list.getBegin();
+		while (true)
 		{
-			m_list.get(iterator->second)->data = std::pair<Key, Value>(k, std::move(v));
-			m_list.moveToFront(iterator->second);
+			std::cout << node->data.first << ' ' << node->data.second << std::endl;
+			if (node->next == invalid_iterator)
+				break;
+			node = m_list.get(node->next);
 		}
 	}
-	//get the value from the cache, or null if key doesnt exist in cache or has been removed. Moves the element to the front of the cache.
-	inline Value get(Key& key)
+#endif // _IRR_DEBUG
+
+	//insert an element into the cache, or update an existing one with the same key
+	inline void insert(Key&& k, Value&& v) { common_insert(std::move(k), std::move(v)); }
+	inline void insert(Key&& k, const Value& v) { common_insert(std::move(k), v); }
+	inline void insert(const Key& k, Value&& v) { common_insert(k, std::move(v)); }
+	inline void insert(const Key& k, const Value& v) { common_insert(k, v); }
+
+	//get the value from cache at an associated Key, or nullptr if Key is not contained within cache. Marks the returned value as most recently used
+	inline Value* get(Key& key)
 	{
 		auto i = common_peek(key);
-		if (i != invalid_iterator)
-		{
+		if (i == invalid_iterator) return nullptr;
+		else {
 			m_list.moveToFront(i);
 			return m_list.get(i)->data.second;
 		}
-		return nullptr;
+	}
+	inline const Value* get(const Key& key) const
+	{
+		return const_cast<base_t*>(this)->get(const_cast<Key&>(key));
 	}
 
-	//get value without reordering the list
-	inline Value peek(Key& key)
+	//get the value from cache at an associated Key, or nullptr if Key is not contained within cache. Does not alter the value use order
+	inline Value* peek(Key& key) const
 	{
 		uint32_t i = common_peek(key);
 		if (i == invalid_iterator) return nullptr;
-		else return  m_list.get(i)->data.second;
+		else return &(m_list.get(i)->data.second);
+	}
+	inline const Value* peek(const Key& key) const
```

Comment:
why are you returning pointers instead of references?

Reply:
to be able to return a nullptr

Reply:
ah ok, for invalid.